File: cpp/2962_count_subarrays/count_subarrays.cpp
#include <algorithm>
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  long long countSubarrays(vector<int>& nums, int k) {
    int mx = *max_element(nums.begin(), nums.end());
    long long ans = 0;
    int cnt = 0, left = 0;
    for (int x : nums) {
      if (x == mx) {
        cnt++;
      }
      while (cnt == k) {
        if (nums[left] == mx) {
          cnt--;
        }
        left++;
      }
      ans += left;
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int, long long>> tests{
      {{1, 3, 2, 3, 3}, 2, 6},
      {{1, 4, 2, 1}, 3, 0},
  };

  for (auto& [nums, k, ans] : tests) {
    assert(Solution().countSubarrays(nums, k) == ans);
  }
}

File: golang/2962_count_subarrays/count_subarrays.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func countSubarrays(nums []int, k int) int64 {
	mx := nums[0]
	for _, v := range nums {
		if v > mx {
			mx = v
		}
	}
	var ans int64 = 0
	cnt := 0
	left := 0
	for _, x := range nums {
		if x == mx {
			cnt++
		}
		for cnt == k {
			if nums[left] == mx {
				cnt--
			}
			left++
		}
		ans += int64(left)
	}
	return ans
}

func main() {
	tests := []struct {
		nums []int
		k    int
		ans  int64
	}{
		{[]int{1, 3, 2, 3, 3}, 2, 6},
		{[]int{1, 4, 2, 1}, 3, 0},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, countSubarrays(test.nums, test.k), index)
	}
}
